packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -0,0 +1,338 @@
+    require(systemAlreadyInitialized(), "Epoch system not initialized");
+    require(isTimeForNextEpoch(), "Epoch is not ready to start");
+    require(!isOnEpochProcess(), "Epoch process is already started");
+    epochProcessing.status = EpochProcessStatus.Started;
+
+    epochs[currentEpochNumber].rewardsBlock = block.number;
+
+    // calculate rewards
+    getEpochRewards().updateTargetVotingYield();
+
+    (
+      uint256 perValidatorReward,
+      uint256 totalRewardsVoter,
+      uint256 totalRewardsCommunity,
+      uint256 totalRewardsCarbonFund
+    ) = getEpochRewards().calculateTargetEpochRewards();
+
+    epochProcessing.perValidatorReward = perValidatorReward;
+    epochProcessing.totalRewardsVoter = totalRewardsVoter;
+    epochProcessing.totalRewardsCommunity = totalRewardsCommunity;
+    epochProcessing.totalRewardsCarbonFund = totalRewardsCarbonFund;
+
+    allocateValidatorsRewards();
+
+    emit EpochProcessingStarted(currentEpochNumber);
+  }
+
+  function finishNextEpochProcess(
+    address[] calldata groups,
+    address[] calldata lessers,
+    address[] calldata greaters
+  ) external nonReentrant {
+    require(isOnEpochProcess(), "Epoch process is not started");
+    // finalize epoch
+    // last block should be the block before and timestamp from previous block
+    epochs[currentEpochNumber].lastBlock = block.number - 1;
+    // start new epoch
+    currentEpochNumber++;
+    epochs[currentEpochNumber].firstBlock = block.number;
+    epochs[currentEpochNumber].startTimestamp = block.timestamp;
+
+    for (uint i = 0; i < elected.length; i++) {
+      address group = getValidators().getValidatorsGroup(elected[i]);
+      if (!processedGroups[group].processed) {
+        epochProcessing.toProcessGroups++;
+        uint256 groupScore = getScoreReader().getGroupScore(group);
+        // We need to precompute epoch rewards for each group since computation depends on total active votes for all groups.
+        uint256 epochRewards = getElection().getGroupEpochRewards(
+          group,
+          epochProcessing.totalRewardsVoter,
+          groupScore
+        );
+        processedGroups[group] = ProcessedGroup(true, epochRewards);
+      }
+    }
+
+    require(epochProcessing.toProcessGroups == groups.length, "number of groups does not match");
+
+    // since we are adding values it makes sense to start from the end
+    for (uint ii = groups.length; ii > 0; ii--) {
+      uint256 i = ii - 1;
+      ProcessedGroup storage processedGroup = processedGroups[groups[i]];
+      // checks that group is actually from elected group
+      require(processedGroup.processed, "group not processed");
+      getElection().distributeEpochRewards(
+        groups[i],
+        processedGroup.epochRewards,
+        lessers[i],
+        greaters[i]
+      );
+
+      // by doing this, we avoid processing a group twice
+      delete processedGroups[groups[i]];
+    }
+    getCeloUnreleasedTreasure().release(
```

Comment:
this and next call could happen on the "start()" epoch process method, and with that we save ourselves from keeping the value on `epochProcessing`

Reply:
doing so would change how much is to be minted, so if done then should be done before  `getEpochRewards().calculateTargetEpochRewards()`, and thus ` getEpochRewards().calculateTargetEpochRewards()` would have to be called twice in start function

Reply:
not really. `calculateTargetEpochRewards()` is called on `startNextEpochProcess()`.

My suggestion was that just after that call, do the `release()` to the community fund and carbon fund, instead of saving those on storage variables and doing them here.

Reply:
We could move the release() call to the community fund and carbon fund in the `startEpochProcess()`, but that means there would be no straight forward way to fetch what was distributed to the community or carbon fund while the epoch is being processed. 

This generally wouldn't be an issue, but since the epoch processing is split into 2 functions that are likely to be called on separate blocks, I think it would be good to receive all the processing information when calling `getEpochProcessingState()` while an epoch is still being processed.